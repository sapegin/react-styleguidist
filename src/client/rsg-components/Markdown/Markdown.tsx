import React, { FC, isValidElement } from 'react';
import PropTypes from 'prop-types';
import { compiler } from 'markdown-to-jsx';
import Link from 'rsg-components/Link';
import Text from 'rsg-components/Text';
import Para from 'rsg-components/Para';
import Code from 'rsg-components/Code';
import MdxBlockquote from 'rsg-components/mdx/MdxBlockquote';
import MdxCodeStatic from 'rsg-components/mdx/MdxCodeStatic';
import MdxHeading from 'rsg-components/mdx/MdxHeading';
import MdxHighlight from 'rsg-components/mdx/MdxHighlight';
import MdxHr from 'rsg-components/mdx/MdxHr';
import MdxList from 'rsg-components/mdx/MdxList';
import MdxPre from 'rsg-components/mdx/MdxPre';
import MdxTable from 'rsg-components/mdx/MdxTable';
import MdxTableBody from 'rsg-components/mdx/MdxTableBody';
import MdxTableCell from 'rsg-components/mdx/MdxTableCell';
import MdxTableHead from 'rsg-components/mdx/MdxTableHead';
import MdxTableRow from 'rsg-components/mdx/MdxTableRow';

// HACK: markdown-to-jsx doesn't destinguish between inline and block code blocks,
// so we highlight syntax in the <pre> tag and assume <code> is always inline.
const Pre: FC = (props) => {
	if (isValidElement(props.children)) {
		return (
			<MdxPre>
				<MdxCodeStatic>
					<MdxHighlight {...props.children.props} />
				</MdxCodeStatic>
			</MdxPre>
		);
	}

	return <MdxPre {...props} />;
};
Pre.propTypes = {
	children: PropTypes.node,
};

export const overrides = {
	a: {
		component: Link,
	},
	h1: {
		component: MdxHeading,
		props: {
			level: 1,
		},
	},
	h2: {
		component: MdxHeading,
		props: {
			level: 2,
		},
	},
	h3: {
		component: MdxHeading,
		props: {
			level: 3,
		},
	},
	h4: {
		component: MdxHeading,
		props: {
			level: 4,
		},
	},
	h5: {
		component: MdxHeading,
		props: {
			level: 5,
		},
	},
	h6: {
		component: MdxHeading,
		props: {
			level: 6,
		},
	},
	p: {
		component: Para,
		props: {
			semantic: 'p',
		},
	},
	em: {
		component: Text,
		props: {
			semantic: 'em',
		},
	},
	strong: {
		component: Text,
		props: {
			semantic: 'strong',
		},
	},
	ul: {
		component: MdxList,
	},
	ol: {
		component: MdxList,
		props: {
			ordered: true,
		},
	},
	blockquote: {
		component: MdxBlockquote,
	},
	code: {
		component: Code,
	},
	pre: {
		component: Pre,
	},
	hr: {
		component: MdxHr,
	},
	table: {
		component: MdxTable,
	},
	thead: {
		component: MdxTableHead,
	},
	th: {
		component: MdxTableCell,
		props: {
			header: true,
		},
	},
	tbody: {
		component: MdxTableBody,
	},
	tr: {
		component: MdxTableRow,
	},
	td: {
		component: MdxTableCell,
	},
} as const;

interface MarkdownProps {
	text: string;
	inline?: boolean;
}

export const Markdown: React.FunctionComponent<MarkdownProps> = ({ text, inline }) => {
	return compiler(text, {
		overrides,
		forceBlock: !inline,
		forceWrapper: inline,
		wrapper: inline ? 'span' : 'div',
	});
};

Markdown.propTypes = {
	text: PropTypes.string.isRequired,
	inline: PropTypes.bool,
};

export default Markdown;
