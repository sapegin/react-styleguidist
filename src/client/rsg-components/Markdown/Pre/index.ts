export { default, PreProps } from 'rsg-components/Markdown/Pre/PreRenderer';
