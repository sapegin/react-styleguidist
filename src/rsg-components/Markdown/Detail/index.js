export { default } from 'rsg-components/Markdown/Detail/DetailRenderer';
export { default as DetailSummary } from 'rsg-components/Markdown/Detail/DetailSummaryRenderer';
