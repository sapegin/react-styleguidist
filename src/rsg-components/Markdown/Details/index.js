export { default } from 'rsg-components/Markdown/Details/DetailsRenderer';
