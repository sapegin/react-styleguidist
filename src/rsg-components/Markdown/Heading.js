import React from 'react';
import PropTypes from 'prop-types';

import Styled from 'rsg-components/Styled';

import HeadingRenderer from 'rsg-components/Heading/HeadingRenderer';

function MarkdownHeading({ classes, depth, children }) {
	return (
		<div className={classes.spacing}>
			<HeadingRenderer depth={depth}>{children}</HeadingRenderer>
		</div>
	);
}

const styles = ({ space }) => ({
	spacing: {
		marginTop: 0,
		marginBottom: space[2],
	},
});

MarkdownHeading.propTypes = {
	classes: PropTypes.object.isRequired,
	depth: PropTypes.oneOf([1, 2, 3, 4, 5, 6]).isRequired,
	children: PropTypes.node,
};

export default Styled(styles)(MarkdownHeading);
