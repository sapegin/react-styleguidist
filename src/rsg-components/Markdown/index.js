import { asArrayMarkdown } from './Markdown';

export { default, asArrayMarkdown } from './Markdown.js';

