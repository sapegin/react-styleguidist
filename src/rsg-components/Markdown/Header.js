import React from 'react';
import PropTypes from 'prop-types';

import Styled from 'rsg-components/Styled';
import { styles as paraStyles } from 'rsg-components/Para';

const styles = ({ space, color, fontSize, fontFamily }) => ({
	para: paraStyles({ space, color, fontFamily }).para,
	h1: {
		composes: '$para',
		fontSize: fontSize.h1,
		fontWeight: 'normal',
	},
	h2: {
		composes: '$para',
		fontSize: fontSize.h2,
		fontWeight: 'normal',
	},
	h3: {
		composes: '$para',
		fontSize: fontSize.h3,
		fontWeight: 'normal',
	},
	h4: {
		composes: '$para',
		fontSize: fontSize.h4,
		fontWeight: 'normal',
	},
	h5: {
		composes: '$para',
		fontSize: fontSize.h5,
		fontWeight: 'normal',
	},
	h6: {
		composes: '$para',
		fontSize: fontSize.h6,
		fontWeight: 'normal',
		fontStyle: 'italic',
	},
});

function Header({ level, classes, children, ...props }) {
	return React.createElement(
		level,
		{
			...props,
			className: classes[level],
		},
		children
	);
}

Header.propTypes = {
	level: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']),
	classes: PropTypes.object.isRequired,
	children: PropTypes.node,
};

const StyledHeader = Styled(styles)(Header);

export default {
	h1: {
		component: StyledHeader,
		props: {
			level: 'h1',
		},
	},
	h2: {
		component: StyledHeader,
		props: {
			level: 'h2',
		},
	},
	h3: {
		component: StyledHeader,
		props: {
			level: 'h3',
		},
	},
	h4: {
		component: StyledHeader,
		props: {
			level: 'h4',
		},
	},
	h5: {
		component: StyledHeader,
		props: {
			level: 'h5',
		},
	},
	h6: {
		component: StyledHeader,
		props: {
			level: 'h6',
		},
	},
};
