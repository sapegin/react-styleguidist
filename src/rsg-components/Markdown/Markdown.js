import React from 'react';
import PropTypes from 'prop-types';
import { compiler } from 'markdown-to-jsx';
import mapValues from 'lodash/mapValues';
// import memoize from 'lodash/memoize';
import Styled from 'rsg-components/Styled';
import Link from 'rsg-components/Link';
import { styles as paraStyles } from 'rsg-components/Para';
import MarkdownHeading from 'rsg-components/Markdown/MarkdownHeading';

// We’re explicitly specifying Webpack loaders here so we could skip specifying them in Webpack configuration.
// That way we could avoid clashes between our loaders and user loaders.
// eslint-disable-next-line import/no-unresolved
require('!!../../../loaders/style-loader!../../../loaders/css-loader!highlight.js/styles/tomorrow.css');

// Temporary disable memoization to fix: https://github.com/styleguidist/react-styleguidist/issues/348
// TODO: Remove after merge: https://github.com/probablyup/markdown-to-jsx/pull/96
const memoize = a => a;

// Code blocks with server-side syntax highlight
function Code({ children, className }) {
	const isHighlighted = className && className.indexOf('lang-') !== -1;
	if (isHighlighted) {
		return <code className={className} dangerouslySetInnerHTML={{ __html: children }} />;
	}
	return <code className={className}>{children}</code>;
}
Code.propTypes = {
	children: PropTypes.node,
	className: PropTypes.string,
};

// Custom CSS classes for each tag: <em> → <em className={s.em}> + custom components
const getBaseOverrides = memoize(classes => {
	const styleOverrides = mapValues(classes, value => ({
		props: {
			className: value,
		},
	}));

	return {
		...styleOverrides,
		a: {
			component: Link,
		},
		h1: {
			component: MarkdownHeading,
			props: {
				level: 1,
			},
		},
		h2: {
			component: MarkdownHeading,
			props: {
				level: 2,
			},
		},
		h3: {
			component: MarkdownHeading,
			props: {
				level: 3,
			},
		},
		h4: {
			component: MarkdownHeading,
			props: {
				level: 4,
			},
		},
		h5: {
			component: MarkdownHeading,
			props: {
				level: 5,
			},
		},
		h6: {
			component: MarkdownHeading,
			props: {
				level: 6,
			},
		},
		code: {
			component: Code,
			props: {
				className: classes.code,
			},
		},
	};
}, () => 'getBaseOverrides');

// Inline mode: replace <p> (usual root component) with <span>
const getInlineOverrides = memoize(classes => {
	const overrides = getBaseOverrides(classes);

	return {
		...overrides,
		p: {
			component: 'span',
			props: {
				className: classes.base,
			},
		},
	};
}, () => 'getInlineOverrides');

const styles = ({ space, fontFamily, fontSize, color, borderRadius }) => ({
	base: {
		color: color.base,
		fontFamily: fontFamily.base,
		fontSize: 'inherit',
	},
	para: paraStyles({ space, color, fontFamily }).para,
	p: {
		composes: '$para',
	},
	ul: {
		composes: '$para',
		paddingLeft: space[3],
	},
	ol: {
		composes: '$para',
		listStyleType: 'decimal',
		paddingLeft: space[3],
	},
	li: {
		composes: '$base',
		listStyleType: 'inherit',
	},
	input: {
		color: color.base,
		display: 'inline-block',
		margin: [[0, '0.35em', '0.25em', '-1.2em']],
		verticalAlign: 'middle',
	},
	blockquote: {
		composes: '$para',
		fontSize: fontSize.base,
		margin: [[space[2], space[4]]],
		padding: 0,
	},
	hr: {
		composes: '$para',
		borderWidth: [[0, 0, 1, 0]],
		borderColor: color.border,
		borderStyle: 'solid',
	},
	em: {
		composes: '$base',
		fontStyle: 'italic',
	},
	strong: {
		composes: '$base',
		fontWeight: 'bold',
	},
	code: {
		fontFamily: fontFamily.monospace,
		fontSize: 'inherit',
		color: 'inherit',
		background: 'transparent',
		whiteSpace: 'inherit',
	},
	pre: {
		composes: '$para',
		backgroundColor: color.codeBackground,
		border: [[1, color.border, 'solid']],
		padding: [[space[1], space[2]]],
		fontSize: fontSize.small,
		borderRadius,
		whiteSpace: 'pre',
	},
	table: {
		composes: '$para',
		borderCollapse: 'collapse',
	},
	thead: {
		composes: '$hr',
	},
	tbody: {},
	td: {
		fontFamily: fontFamily.base,
		padding: [[space[0], space[2], space[0], 0]],
		fontSize: fontSize.base,
	},
	th: {
		composes: '$td',
		fontWeight: 'bold',
	},
	tr: {},
});

function Markdown({ classes, text, inline }) {
	const overrides = inline ? getInlineOverrides(classes) : getBaseOverrides(classes);
	return compiler(text, { overrides });
}

Markdown.propTypes = {
	classes: PropTypes.object.isRequired,
	text: PropTypes.string.isRequired,
	inline: PropTypes.bool,
};

export default Styled(styles)(Markdown);
