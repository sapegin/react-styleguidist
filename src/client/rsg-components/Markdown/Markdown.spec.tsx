import React from 'react';
import { render, screen } from '@testing-library/react';
import Markdown from './Markdown';

const normalizeClassNames = (html: string): string => html.replace(/(rsg--\w+)-\d+/g, '$1-0');

describe('Markdown', () => {
	test.each([
		['links', 'a [link](http://test.com)'],
		[
			'paragraphs',
			`
a paragraph

another paragraph
		`,
		],
		[
			'unordered lists',
			`
* list
* item
* three
			`,
		],
		[
			'ordered lists',
			`
1. list
1. item
1. three
			`,
		],
		[
			'mixed nested lists',
			`
* list 1
* list 2
	1. Sub-list
	1. Sub-list
	1. Sub-list
* list 3
			`,
		],
		[
			'blockquotes',
			`
> This is a blockquote.
> And this is a second line.
			`,
		],
		[
			'emphasis and strong text',
			`
this text is **strong**

and this is _emphasized_
			`,
		],
		[
			'headings with generated IDs',
			`
# one
## two
### three
#### four
##### five
###### six
			`,
		],
		[
			'pre-formatted text',
			`
    this is preformatted
    so is this
		`,
		],
		[
			'code blocks with highlighting',
			`
\`\`\`html
<foo></foo>
\`\`\`
		`,
		],
		['inline code with escaping', 'Foo `<bar>` baz'],
		['horizontal rules', `---`],
		[
			'tables',
			`
| heading 1 | heading 2 |
| --------- | --------- |
| foo		| bar		|
| more foo	| more bar	|
		`,
		],
		[
			// TODO: fix <input> styles
			'check-lists',
			`
* [ ] to do 1
* [ ] to do 2
* [x] to do 3
			`,
		],
	])('renders %s', (_, markdown) => {
		const { container } = render(<Markdown text={markdown} />);
		expect(normalizeClassNames(container.innerHTML)).toMatchSnapshot();
	});

	test('should forward DOM attributes onto resulting HTML', () => {
		const markdown =
			'<a href="test.com" id="preserve-my-id" class="preserve-my-class">Something</a>';

		const { container } = render(<Markdown text={markdown} />);

		expect(container.querySelector('a')?.id).toEqual('preserve-my-id');
		expect(container.querySelector('a')?.className).toContain('preserve-my-class');
	});

	test('should ignore single line comments', () => {
		const markdown = `Hello World
<!-- This is a single line comment -->
`;
		render(<Markdown text={markdown} />);

		expect(screen.queryByText(/comment/i)).not.toBeInTheDocument();
	});

	test('should ignore multiline comments', () => {
		const markdown = `Hello World
<!--
This is a
multiline
comment
-->
`;
		render(<Markdown text={markdown} />);

		expect(screen.queryByText(/comment/i)).not.toBeInTheDocument();
	});
});

describe('Markdown inline', () => {
	test('should render text in a span', () => {
		const { container } = render(<Markdown text="Hello world!" inline />);
		expect(normalizeClassNames(container.innerHTML)).toMatchSnapshot();
	});
});
