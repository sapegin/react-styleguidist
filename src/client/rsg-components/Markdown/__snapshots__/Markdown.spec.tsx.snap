// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown inline should render text in a span 1`] = `
<span>
  Hello world!
</span>
`;

exports[`Markdown renders blockquotes 1`] = `
<blockquote class="rsg--blockquote-0">
  <p class="rsg--para-0">
    This is a blockquote.
And this is a second line.
  </p>
</blockquote>
`;

exports[`Markdown renders check-lists 1`] = `
<ul class="rsg--list-0">
  <li class="rsg--li-0">
    <input readonly
           type="checkbox"
    >
    to do 1
  </li>
  <li class="rsg--li-0">
    <input readonly
           type="checkbox"
    >
    to do 2
  </li>
  <li class="rsg--li-0">
    <input readonly
           type="checkbox"
           checked
    >
    to do 3
  </li>
</ul>
`;

exports[`Markdown renders code blocks with highlighting 1`] = `
<div class="rsg--preContainer-0">
  <div class="rsg--container-0">
    <pre class="prism-code language-html rsg--highlightPre-0"
         style="color: rgb(51, 51, 51); background-color: transparent;"
    >
      <div class="token-line"
           style="color: rgb(51, 51, 51);"
      >
        <span class="token tag punctuation"
              style="color: rgb(153, 153, 153);"
        >
          &lt;
        </span>
        <span class="token tag"
              style="color: rgb(153, 0, 85);"
        >
          foo
        </span>
        <span class="token tag punctuation"
              style="color: rgb(153, 153, 153);"
        >
          &gt;
        </span>
        <span class="token tag punctuation"
              style="color: rgb(153, 153, 153);"
        >
          &lt;/
        </span>
        <span class="token tag"
              style="color: rgb(153, 0, 85);"
        >
          foo
        </span>
        <span class="token tag punctuation"
              style="color: rgb(153, 153, 153);"
        >
          &gt;
        </span>
      </div>
    </pre>
  </div>
</div>
`;

exports[`Markdown renders emphasis and strong text 1`] = `
<div>
  <p class="rsg--para-0">
    this text is
    <strong class="rsg--text-0 rsg--inheritSize-0 rsg--baseColor-0 rsg--strong-0">
      strong
    </strong>
  </p>
  <p class="rsg--para-0">
    and this is
    <em class="rsg--text-0 rsg--inheritSize-0 rsg--baseColor-0 rsg--em-0">
      emphasized
    </em>
  </p>
</div>
`;

exports[`Markdown renders headings with generated IDs 1`] = `
<div>
  <h1 id="one"
      class="rsg--heading-0 rsg--heading1-0 rsg--spacing-0"
  >
    one
  </h1>
  <h2 id="two"
      class="rsg--heading-0 rsg--heading2-0 rsg--spacing-0"
  >
    two
  </h2>
  <h3 id="three"
      class="rsg--heading-0 rsg--heading3-0 rsg--spacing-0"
  >
    three
  </h3>
  <h4 id="four"
      class="rsg--heading-0 rsg--heading4-0 rsg--spacing-0"
  >
    four
  </h4>
  <h5 id="five"
      class="rsg--heading-0 rsg--heading5-0 rsg--spacing-0"
  >
    five
  </h5>
  <h6 id="six"
      class="rsg--heading-0 rsg--heading6-0 rsg--spacing-0"
  >
    six
  </h6>
</div>
`;

exports[`Markdown renders horizontal rules 1`] = `<hr class="rsg--hr-0">`;

exports[`Markdown renders inline code with escaping 1`] = `
<p class="rsg--para-0">
  Foo
  <code class="rsg--code-0">
    &lt;bar&gt;
  </code>
  baz
</p>
`;

exports[`Markdown renders links 1`] = `
<p class="rsg--para-0">
  a
  <a href="http://test.com"
     class="rsg--link-0"
  >
    link
  </a>
</p>
`;

exports[`Markdown renders mixed nested lists 1`] = `
<ul class="rsg--list-0">
  <li class="rsg--li-0">
    list 1
  </li>
  <li class="rsg--li-0">
    list 2
    <ol class="rsg--list-0 rsg--ordered-0">
      <li class="rsg--li-0">
        Sub-list
      </li>
      <li class="rsg--li-0">
        Sub-list
      </li>
      <li class="rsg--li-0">
        Sub-list
      </li>
    </ol>
  </li>
  <li class="rsg--li-0">
    list 3
  </li>
</ul>
`;

exports[`Markdown renders ordered lists 1`] = `
<ol class="rsg--list-0 rsg--ordered-0">
  <li class="rsg--li-0">
    list
  </li>
  <li class="rsg--li-0">
    item
  </li>
  <li class="rsg--li-0">
    three
  </li>
</ol>
`;

exports[`Markdown renders paragraphs 1`] = `
<div>
  <p class="rsg--para-0">
    a paragraph
  </p>
  <p class="rsg--para-0">
    another paragraph
  </p>
</div>
`;

exports[`Markdown renders pre-formatted text 1`] = `
<div class="rsg--preContainer-0">
  <div class="rsg--container-0">
    <pre class="rsg--highlightPre-0">
      this is preformatted
so is this
    </pre>
  </div>
</div>
`;

exports[`Markdown renders tables 1`] = `
<table class="rsg--table-0">
  <thead class="rsg--thead-0">
    <tr>
      <th class="rsg--th-0 rsg--td-0">
        heading 1
      </th>
      <th class="rsg--th-0 rsg--td-0">
        heading 2
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td class="rsg--td-0">
        foo
      </td>
      <td class="rsg--td-0">
        bar
      </td>
    </tr>
    <tr>
      <td class="rsg--td-0">
        more foo
      </td>
      <td class="rsg--td-0">
        more bar
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Markdown renders unordered lists 1`] = `
<ul class="rsg--list-0">
  <li class="rsg--li-0">
    list
  </li>
  <li class="rsg--li-0">
    item
  </li>
  <li class="rsg--li-0">
    three
  </li>
</ul>
`;
